README.md
# gacp

[![Build Status][travis-image]][travis-url]
[![NPM Version][npm-version-image]][npm-url]
[![NPM Downloads][npm-downloads-image]][npm-url]
[![MIT License][license-image]][license-url]
[![Standard Version][standard-version-image]][standard-version-url]
[![Codecov][codecov-image]][codecov-url]
[![Financial Contributors on Open Collective](https://opencollective.com/gacp/all/badge.svg?label=financial+contributors)](https://opencollective.com/gacp)

💬Git add, commit, push with [Conventional Commits](https://www.conventionalcommits.org/) and [Gitmoji](https://gitmoji.carloscuesta.me/).

![GACP](./assets/images/gacp.gif)

## Installation

`npm i -g gacp`

## Usage

`gacp`

`gacp --help`

`gacp --no-add`

`gacp --no-push`

`gacp --emoji emoji`

`gacp --editor`

`gacp --no-verify`

`gacp --emoji none`

## Configuration File

You can configure gacp via:

- A `gacp` property in `package.json`.
- A `.gacprc` file in JSON, YAML or CommonJS with or without extensions `.json`, `.yaml`, `.yml`, `.js`.
- A `gacp.config.js` file in CommonJS.

### Basic Configuration

Default configuration:

```json
{
  "add": true,
  "push": true,
  "emoji": "code",
  "editor": false,
  "hooks": {
    "postpush": ""
  }
}
```

## Change log

[Change log](CHANGELOG.md)

## Contributing

[Contributing](CONTRIBUTING.md)

## Contributors

### Code Contributors

This project exists thanks to all the people who contribute. [[Contribute](CONTRIBUTING.md)].
<a href="https://github.com/vivaxy/gacp/graphs/contributors"><img src="https://opencollective.com/gacp/contributors.svg?width=890&button=false" /></a>

### Financial Contributors

Become a financial contributor and help us sustain our community. [[Contribute](https://opencollective.com/gacp/contribute)]

#### Individuals

<a href="https://opencollective.com/gacp"><img src="https://opencollective.com/gacp/individuals.svg?width=890"></a>

#### Organizations

Support this project with your organization. Your logo will show up here with a link to your website. [[Contribute](https://opencollective.com/gacp/contribute)]

<a href="https://opencollective.com/gacp"><img src="https://opencollective.com/gacp/organization.svg?width=890"></a>

## Related Projects

- [VSCode Conventional Commits](https://github.com/vivaxy/vscode-conventional-commits)
- [gcmt](https://github.com/vivaxy/gcmt)
- [commitizen](https://github.com/commitizen/cz-cli)
- [conventional-changelog](https://github.com/conventional-changelog/conventional-changelog)
- [gitmoji-cli](https://github.com/carloscuesta/gitmoji-cli)
- [git-commit-cp](https://github.com/Dolov/git-commit-cp)
- [git-cz](https://github.com/streamich/git-cz)
- [cz-emoji](https://github.com/ngryman/cz-emoji)

#

[travis-image]: https://img.shields.io/travis/vivaxy/gacp.svg?style=flat-square
[travis-url]: https://travis-ci.org/vivaxy/gacp
[npm-version-image]: http://img.shields.io/npm/v/gacp.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/gacp
[npm-downloads-image]: https://img.shields.io/npm/dt/gacp.svg?style=flat-square
[license-image]: https://img.shields.io/npm/l/gacp.svg?style=flat-square
[license-url]: LICENSE
[standard-version-image]: https://img.shields.io/badge/release-standard%20version-brightgreen.svg?style=flat-square
[standard-version-url]: https://github.com/conventional-changelog/standard-version
[codecov-image]: https://img.shields.io/codecov/c/github/vivaxy/gacp.svg?style=flat-square
[codecov-url]: https://codecov.io/gh/vivaxy/gacp


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [3.0.3](https://github.com/vivaxy/gacp/compare/v3.0.2...v3.0.3) (2022-08-17)


### Bug Fixes

* :bug:  cannot push when multiple remotes ([002ecbd](https://github.com/vivaxy/gacp/commit/002ecbdef192439646103430b81dd8d6ae15374b)), closes [#92](https://github.com/vivaxy/gacp/issues/92)

### [3.0.2](https://github.com/vivaxy/gacp/compare/v3.0.1...v3.0.2) (2022-05-19)


### Bug Fixes

* :bug:  fix no-verify behaviour ([156c75b](https://github.com/vivaxy/gacp/commit/156c75bb7cc17398b6b285f2bfe22e2b50479d0e)), closes [#124](https://github.com/vivaxy/gacp/issues/124)

### [3.0.1](https://github.com/vivaxy/gacp/compare/v3.0.0...v3.0.1) (2022-05-05)


### Bug Fixes

* :bug:  revert default emoji config from none to code ([3000e9e](https://github.com/vivaxy/gacp/commit/3000e9e79e5caa8fe815ae8d7f4c1cbad4bfe625))

## [3.0.0](https://github.com/vivaxy/gacp/compare/v2.10.2...v3.0.0) (2022-05-04)


### ⚠ BREAKING CHANGES

* set gitmoji disabled by default. closes: #118
* gacp no longer push tags. closes #123

### Features

*  support disable gitmoji ([0b9e9dd](https://github.com/vivaxy/gacp/commit/0b9e9dd048e42e4f2e92c8097c49a78c6f71c4b3))
* **add `--no-verify` option`:** :sparkles: ([1d81dab](https://github.com/vivaxy/gacp/commit/1d81dabcba5a250b15cd99da6abd7293da9d67f4)), closes [#121](https://github.com/vivaxy/gacp/issues/121)
* remove `--follow-tags` when pushing ([f94879e](https://github.com/vivaxy/gacp/commit/f94879e392d00ca000b08cce9e900b2f38e1b0af))


### Bug Fixes

* **deps:** update dependency execa to v5 ([c8d39de](https://github.com/vivaxy/gacp/commit/c8d39defe4d2e81966368817b74d704b8ad577f0))
* **deps:** update dependency fs-extra to v10 ([69d3ace](https://github.com/vivaxy/gacp/commit/69d3acec47a02629943b44b41be3084cdc43cfac))
* **deps:** update dependency yargs to v17 ([78ea990](https://github.com/vivaxy/gacp/commit/78ea9902ff6b198deeb19a0240e81e54cf5c5028))

### [2.10.2](https://github.com/vivaxy/gacp/compare/v2.10.1...v2.10.2) (2020-10-13)


### Bug Fixes

* :bug:  fix history not cleared ([b7fb625](https://github.com/vivaxy/gacp/commit/b7fb625d2aca5fd7b5042d7bc1d885d9e4adde87))
* :bug:  fix history not cleared after a successfully commit ([c06e3e6](https://github.com/vivaxy/gacp/commit/c06e3e6fb07f9e73dd31f1c38fe48d8ba8070fbe))

### [2.10.1](https://github.com/vivaxy/gacp/compare/v2.10.0...v2.10.1) (2020-10-13)


### Bug Fixes

* :bug: fix history not working ([8cb774d](https://github.com/vivaxy/gacp/commit/8cb774d6ff11a40b0a5a6268e2a06a1fe57492ab))
* **deps:** update dependency cosmiconfig to v7 ([6cd362b](https://github.com/vivaxy/gacp/commit/6cd362b1eb54b005337d37af0f9f4801c04ddc78))
* **deps:** update dependency update-notifier to v5 ([781ce71](https://github.com/vivaxy/gacp/commit/781ce71eae1d656a59a2d4c019ff8dcd4902707c))
* **deps:** update dependency yargs to v16 ([98f5753](https://github.com/vivaxy/gacp/commit/98f57538a78e926080b4cd344c5ed95a1cfe0d6f))

## [2.10.0](https://github.com/vivaxy/gacp/compare/v2.9.0...v2.10.0) (2020-06-23)


### Features

* :sparkles:  support multi-line longer description ([42c431a](https://github.com/vivaxy/gacp/commit/42c431a73974b293df730731ffe75ccd46aec1e4)), closes [#40](https://github.com/vivaxy/gacp/issues/40) [#43](https://github.com/vivaxy/gacp/issues/43)


### Bug Fixes

* **deps:** update dependency chalk to v4 ([4dd27bd](https://github.com/vivaxy/gacp/commit/4dd27bd040488599c57089fb74fe472b2188b2a7))
* **deps:** update dependency conventional-commit-types to v3 ([44fe77d](https://github.com/vivaxy/gacp/commit/44fe77d06b988447d93d6e2d0f8ef31ef869cf6e))
* **deps:** update dependency cosmiconfig to v6 ([bf0168e](https://github.com/vivaxy/gacp/commit/bf0168e555953f412d3597bebbe2a3244c31d0b2))
* **deps:** update dependency execa to ^0.11.0 ([55e4b14](https://github.com/vivaxy/gacp/commit/55e4b147e640ae137151c39f465d094a9d255cce))
* **deps:** update dependency execa to v4 ([87a4dc4](https://github.com/vivaxy/gacp/commit/87a4dc48269ec443e259b3fa3dffcdf20036ec21))
* **deps:** update dependency fs-extra to v9 ([ada4c09](https://github.com/vivaxy/gacp/commit/ada4c092147eb0c65fa003d84f5808d0713345b5))
* **deps:** update dependency update-notifier to v4 ([2e1670f](https://github.com/vivaxy/gacp/commit/2e1670f3df1ae59ced7b06c390ee349a1d68298a))
* **deps:** update dependency yargs to v15 ([f8cde91](https://github.com/vivaxy/gacp/commit/f8cde91994068a23d5c4f77b1d450274c6a73059))

## [2.9.0](https://github.com/vivaxy/gacp/compare/v2.8.0...v2.9.0) (2020-05-05)


### Features

* :sparkles:  support run `gacp` in a git subdirectory ([9bad872](https://github.com/vivaxy/gacp/commit/9bad872))



## [2.8.0](https://github.com/vivaxy/gacp/compare/v2.7.2...v2.8.0) (2020-04-29)


### Bug Fixes

* :bug: fix commitlint rule when rule config is set to `never` ([6ffbe7c](https://github.com/vivaxy/gacp/commit/6ffbe7c))


### Build System

* :children_crossing:  add vscode debug config ([2aa6aaa](https://github.com/vivaxy/gacp/commit/2aa6aaa))


### Features

* :sparkles:  support run postpush hook ([d7bc7d8](https://github.com/vivaxy/gacp/commit/d7bc7d8)), closes [#45](https://github.com/vivaxy/gacp/issues/45) [#46](https://github.com/vivaxy/gacp/issues/46)
* :sparkles: add `--set-upstream` option when `git push` ([50d9258](https://github.com/vivaxy/gacp/commit/50d9258)), closes [#44](https://github.com/vivaxy/gacp/issues/44)



### [2.7.2](https://github.com/vivaxy/gacp/compare/v2.7.1...v2.7.2) (2020-02-25)


### Bug Fixes

* **typo:** :rewind:  revert last fix, `terse` to `tense` ([5aa6bc1](https://github.com/vivaxy/gacp/commit/5aa6bc1))



### [2.7.1](https://github.com/vivaxy/gacp/compare/v2.7.0...v2.7.1) (2020-02-25)


### Bug Fixes


src/bin.ts
#!/usr/bin/env node
/**
 * @since 2016-11-22 13:29
 * @author vivaxy
 */
import * as yargs from 'yargs';
import chalk = require('chalk');
import * as log from 'log-util';
import { cosmiconfig } from 'cosmiconfig';
import * as updateNotifier from 'update-notifier';

import gacp from './gacp';
import { flushHistory } from './messages/history';
import { EMOJI_TYPES } from './configs';

const pkg = require('../package.json');

function debug(...message: any[]) {
  log.debug('gacp:bin', ...message);
}

async function configureYargs() {
  const explorer = cosmiconfig(pkg.name);
  const cosmiconfigResult = await explorer.search();
  const { config = {} } = cosmiconfigResult || {};
  ((yargs
    .options({
      add: {
        type: 'boolean',
        desc: 'run git add .',
        default: true,
      },
      push: {
        type: 'boolean',
        desc: 'run git push',
        default: true,
      },
      emoji: {
        type: 'string',
        desc: 'use emoji or code',
        choices: ['none', 'code', 'emoji'],
        default: 'code',
      },
      editor: {
        type: 'boolean',
        desc: 'use node external editor in longer description',
        default: false,
      },
      logLevel: {
        type: 'number',
        desc: 'log level',
        default: 1,
      },
      cwd: {
        type: 'string',
        desc: 'working directory',
        default: process.cwd(),
      },
      verify: {
        type: 'boolean',
        desc: 'verify git commit and git push',
        default: true,
      },
    })
    .config(config)
    .help()
    .version().argv as unknown) as any)._;

  const hooks = config.hooks || {};
  return {
    hooks: {
      postpush: hooks.postpush || '',
    },
  };
}

function notifyUpdate() {
  const notifier = updateNotifier({ pkg });
  const { update } = notifier;
  if (update) {
    const installCommand = `npm i -g ${update.name}`;
    const message =
      'Update available ' +
      chalk.dim(update.current) +
      chalk.reset(' → ') +
      chalk.green(update.latest) +
      ' \nRun ' +
      chalk.cyan(installCommand) +
      ' to update';

    notifier.notify({ message });
  }
}

(async function () {
  try {
    notifyUpdate();
    const extraConfigs = await configureYargs();
    const {
      logLevel,
      cwd,
      emoji,
      add,
      push,
      verify,
      editor,
    } = (yargs.argv as unknown) as {
      logLevel: number;
      cwd: string;
      emoji: EMOJI_TYPES;
      add: boolean;
      push: boolean;
      verify: boolean;
      editor: boolean;
    };
    log.setLevel(logLevel);
    await gacp({
      cwd: cwd as string,
      add: add as boolean,
      push: push as boolean,
      emoji: emoji as EMOJI_TYPES,
      editor: editor as boolean,
      verify: verify as boolean,
      hooks: {
        postpush: extraConfigs.hooks.postpush,
      },
    });
    await flushHistory();
  } catch (e) {
    log.error(e.message);
    log.debug(e.stack);
    await flushHistory();
    process.exit(1);
  }
})();
